packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -1824,9 +1824,106 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
   return root;
 }
 
-export function warnIfNotCurrentlyBatchingInDev(fiber: Fiber): void {
+// in a test-like environment, we want to warn if dispatchAction() is
+// called outside of a TestUtils.act(...)/batchedUpdates/render call.
+// so we have a a step counter for when we descend/ascend from
+// actedUpdates() calls, and test on it for when to warn
+let actingUpdatesScopeDepth = 0;
+
+// we 'wait' for queued microtasks to resolve (after flushing effects)
+// using either setImmediate when available, or postMessage
+const nextTick =
+  typeof setImmediate !== 'undefined'
+    ? setImmediate
+    : callback => {
+        const channel = new MessageChannel();
+        channel.port1.onmessage = callback;
+        channel.port2.postMessage(undefined);
+      };
+
+export function actedUpdates(callback: () => void | Promise<void>) {
+  let previousActingUpdatesScopeDepth;
   if (__DEV__) {
-    if (isRendering === false && isBatchingUpdates === false) {
+    previousActingUpdatesScopeDepth = actingUpdatesScopeDepth;
+    actingUpdatesScopeDepth++;
+  }
+
+  function warnIfScopeDepthMismatch() {
+    if (__DEV__) {
+      if (actingUpdatesScopeDepth > previousActingUpdatesScopeDepth) {
+        // if it's _less than_ previousActingUpdatesScopeDepth, then we can assume the 'other' one has warned
+        warningWithoutStack(
+          null,
+          'You seem to have overlapping act() calls, this is not supported. ' +
+            'Be sure to await previous act() calls before making a new one. ',
+        );
+      }
+    }
+  }
+
+  function flushEffectsAndMicroTasks() {
+    // eslint-disable-next-line no-undef
+    return new Promise((resolve, reject) => {
```

Comment:
Use a thenable instead of a promise

Reply:
I tried, but it still needs the microtask before running flushPassiveEffects again, making this hairier. I'm moving this out of ReactFiberScheduler (into TestUtils/TestRenderer/createReactNoop), so it's... fine, maybe?

Reply:
consider also that this codepath is only called with the async version of act, which means they probably have Promise defined on the page